File: app/src/main/java/be/digitalia/fosdem/viewmodels/EventDetailsViewModel.java
package be.digitalia.fosdem.viewmodels;

import android.app.Application;
import androidx.annotation.NonNull;
import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.Transformations;
import be.digitalia.fosdem.db.AppDatabase;
import be.digitalia.fosdem.model.Event;
import be.digitalia.fosdem.model.EventDetails;

public class EventDetailsViewModel extends AndroidViewModel {

	private final AppDatabase appDatabase = AppDatabase.getInstance(getApplication());
	private final MutableLiveData<Event> event = new MutableLiveData<>();
	private final LiveData<EventDetails> eventDetails = Transformations.switchMap(event,
			event -> appDatabase.getScheduleDao().getEventDetails(event));

	public EventDetailsViewModel(@NonNull Application application) {
		super(application);
	}

	public void setEvent(@NonNull Event event) {
		if (!event.equals(this.event.getValue())) {
			this.event.setValue(event);
		}
	}

	public LiveData<EventDetails> getEventDetails() {
		return eventDetails;
	}
}

File: app/src/main/kotlin/be/digitalia/fosdem/viewmodels/EventDetailsViewModel.kt
package be.digitalia.fosdem.viewmodels

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import be.digitalia.fosdem.db.ScheduleDao
import be.digitalia.fosdem.model.Event
import be.digitalia.fosdem.model.EventDetails
import dagger.assisted.Assisted
import dagger.assisted.AssistedFactory
import dagger.assisted.AssistedInject
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.Deferred
import kotlinx.coroutines.async

@HiltViewModel(assistedFactory = EventDetailsViewModel.Factory::class)
class EventDetailsViewModel @AssistedInject constructor(
    scheduleDao: ScheduleDao,
    @Assisted event: Event
) : ViewModel() {

    val eventDetails: Deferred<EventDetails> = viewModelScope.async {
        scheduleDao.getEventDetails(event)
    }

    @AssistedFactory
    interface Factory {
        fun create(event: Event): EventDetailsViewModel
    }
}